Add section on using CI artifacts in PRs.

bdice opened the pull request:
Adding a section on this to the docs because I've needed to refer to it when answering questions at least 3 times.

cc: @ajschmidt8 @vyasr @harrism

bdice commented:
Preview: https://deploy-preview-444--docs-rapids-ai.netlify.app/resources/github-actions/#using-ci-artifacts-in-other-prs

vyasr commented on resources/github-actions.md:
```diff
@@ -45,6 +45,59 @@ There is a link provided at the end of every C++ and Python build job where the
+
+For changes that cross library boundaries, it may be necessary to test a pull request to one library with changes from a pull request to another library.
+To do this, it is necessary to download CI artifacts (described in the above section) from one library during the CI workflow of another library.
+First, determine the pull request number and commit hash that need to be tested from the other library.
```
suggestion (non-blocking): It might be helpful to explain how the links described in the above section translate to the arguments to `rapids-get-artifact` below.

vyasr reviewed (approved):
praise: Thanks for adding this! I've definitely had to figure this out on more than one occasion. Everything looks good.

harrism commented on resources/github-actions.md:
```diff
@@ -45,6 +45,59 @@ There is a link provided at the end of every C++ and Python build job where the
+
+Note that the custom channel for PR artifacts is needed in the build scripts _and_ the test scripts.
+If building/testing a Python package that depends on a C++ library, it is also necessary to use PR artifacts from that C++ library (e.g. if testing `rmm` artifacts, use the corresponding `librmm` CI artifacts as well).
+
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion: (non-blocking)** ‏ You could add a section with the tip you gave me to automatically grab the commit hash given the PR: 

```
REPO="rmm"
PR_NUMBER="1095"
COMMIT=$(git ls-remote https://github.com/rapidsai/${REPO}.git refs/heads/pull-request/${PR_NUMBER} | cut -c1-7)
```

bdice replied:
I'll rewrite this with https://github.com/rapidsai/gha-tools/pull/85.

harrism commented on resources/github-actions.md:
````diff
@@ -45,6 +45,59 @@ There is a link provided at the end of every C++ and Python build job where the
+  cudf libcudf
+```
+
+Note that the custom channel for PR artifacts is needed in the build scripts _and_ the test scripts.
````
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ In some repos, the `test_python.sh` is quite complicated with multiple `rapids-mamba-retry` calls. I think you should state that the python and C++ artifact channels need to be added to every call "just in case".

**💡 suggestion:** ‏ It would also be nice to mention the dependency graph and that, for example, if you are testing artifacts from an RMM PR rmm#A in cuML, you probably also need to create a cuDF PR cudf#B that uses the artifacts from rmm#A, and then your cuML test PR will need to include the artifact channels for both rmm#A and cudf#B.

harrism reviewed (changes requested):
<!-- Praise highlights something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
**👏 praise:** ‏ This is awesome and I think it will be a big help. I just have a few small suggestions.

bdice commented:
Re-requesting review from @harrism and @vyasr. I rewrote this to use https://github.com/rapidsai/gha-tools/pull/85 and I think I've addressed your feedback.

ajschmidt8 reviewed (commented):
Looks good. Appreciate you documenting this.

Presumably the changes to the downstream repos should only be temporary, right?

Should we add a sentence or two mentioning that they should be reverted once testing is complete?

ajschmidt8 commented on resources/github-actions.md:
```diff
@@ -45,6 +45,61 @@ There is a link provided at the end of every C++ and Python build job where the
 
 ![](/assets/images/downloads.png)
 
+## Using CI Artifacts in Other PRs
```
```suggestion
## Using Conda CI Artifacts in Other PRs
```

Should we clarify this section header since it only applies to Conda CI artifacts?

harrism commented on resources/github-actions.md:
```diff
@@ -45,6 +45,61 @@ There is a link provided at the end of every C++ and Python build job where the
+
+For changes that cross library boundaries, it may be necessary to test a pull request to one library with changes from a pull request to another library.
+Consider the overall RAPIDS dependency graph when testing.
+For example, if you are testing artifacts from an RMM PR rmm#A in cuML, you probably also need to create a cuDF PR cudf#B that uses the artifacts from rmm#A, and then your cuML test PR will need to include the artifact channels for both rmm#A and cudf#B.
```
<!-- Nitpicks are trivial preference-based requests. These should be non-blocking by nature.  -->
**🤓 nitpick (non-blocking):** ‏ I think it helps readability to put the PR IDs in back quotes.

```suggestion
For example, if you are testing artifacts from an RMM PR `rmm#A` in cuML, you probably also need to create a cuDF PR `cudf#B` that uses the artifacts from `rmm#A`, and then your cuML test PR will need to include the artifact channels for both `rmm#A` and `cudf#B`.
```

harrism commented on resources/github-actions.md:
```diff
@@ -45,6 +45,61 @@ There is a link provided at the end of every C++ and Python build job where the
+Note that the custom channel for PR artifacts is needed in the build scripts _and_ the test scripts, for C++ _and_ Python.
+If building/testing a Python package that depends on a C++ library, it is necessary to use PR artifacts from that C++ library and not just Python (e.g. if testing `rmm` artifacts, you must use the corresponding `librmm` CI artifacts as well as `rmm`).
+In some repos, the `test_python.sh` is quite complicated with multiple calls to conda/mamba.
+We recommend that the Python and C++ artifact channels should be added to every call "just in case."
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ For clarity:

```suggestion
We recommend that the Python and C++ artifact channels should be added to every call of `rapids-mamba-retry`/`rapids-conda-retry` "just in case."
```

harrism commented on resources/github-actions.md:
```diff
@@ -45,6 +45,61 @@ There is a link provided at the end of every C++ and Python build job where the
+If building/testing a Python package that depends on a C++ library, it is necessary to use PR artifacts from that C++ library and not just Python (e.g. if testing `rmm` artifacts, you must use the corresponding `librmm` CI artifacts as well as `rmm`).
+In some repos, the `test_python.sh` is quite complicated with multiple calls to conda/mamba.
+We recommend that the Python and C++ artifact channels should be added to every call "just in case."
+
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ Perhaps add a tip about `rapids-get-pr-conda-artifact`:
```suggestion

Note: By default `rapids-get-pr-conda-artifact` uses the most recent commit from the specified PR. A commit hash from the dependent PR can be added as an optional 4th argument to test with an earlier commit or to pin testing to a commit even if the dependent PR is updated.
```

harrism reviewed (commented):
Thanks for updating! A few additional suggestions.

bdice commented on resources/github-actions.md:
```diff
@@ -45,6 +45,61 @@ There is a link provided at the end of every C++ and Python build job where the
+First, determine the pull request number(s) to be tested from the other library.
+Then, fetch the CI artifacts from the other library's pull request and use them when building and testing.
+The example code below demonstrates building and testing with conda packages from other library PRs.
+Replace the pull request numbers and library names as needed.
```
```suggestion
Replace the pull request numbers and library names as needed.
Remember that changes to use CI artifacts should be _temporary_ and should be reverted prior to merging any required changes in that PR.
```

bdice commented on resources/github-actions.md:
```diff
@@ -45,6 +45,63 @@ There is a link provided at the end of every C++ and Python build job where the
+In some repos, the `test_python.sh` is quite complicated with multiple calls to conda/mamba.
+We recommend that the Python and C++ artifact channels should be added to every call of `rapids-mamba-retry` / `rapids-conda-retry` "just in case."
+
+Note: By default `rapids-get-pr-conda-artifact` uses the most recent commit from the specified PR. A commit hash from the dependent PR can be added as an optional 4th argument to test with an earlier commit or to pin testing to a commit even if the dependent PR is updated.
```
```suggestion
Note: By default `rapids-get-pr-conda-artifact` uses the most recent commit from the specified PR.
A commit hash from the dependent PR can be added as an optional 4th argument to test with an earlier commit or to pin testing to a commit even if the dependent PR is updated.
```

bdice commented:
@ajschmidt8 @harrism I like the suggestions you made, and applied them all!

harrism reviewed (approved):
Great work, and thank you.
